ipa-core/src/protocol/ipa_prf/aggregation/breakdown_reveal.rs

```diff
@@ -0,0 +1,162 @@
+use futures::{stream::Stream, TryStreamExt};
+
+use crate::{
+    error::Error,
+    ff::{
+        boolean::Boolean,
+        boolean_array::{BooleanArray, BA64},
+        ArrayAccess, U128Conversions,
+    },
+    helpers::{repeat_n, TotalRecords},
+    protocol::{
+        basics::Reveal,
+        boolean::step::SixteenBitStep,
+        context::{Context, UpgradedSemiHonestContext},
+        ipa_prf::{
+            aggregation::step::AggregationStep, boolean_ops::addition_sequential::integer_sat_add,
+            prf_sharding::SecretSharedAttributionOutputs, shuffle::shuffle_attribution_outputs,
+            BreakdownKey,
+        },
+        RecordId,
+    },
+    secret_sharing::{replicated::semi_honest::AdditiveShare as Replicated, BitDecomposed},
+    sharding::NotSharded,
+};
+
+// Improved Aggregation a.k.a Aggregation revealing breakdown.
+//
+// The previous phase was attribution. The input to aggregation is a stream of
+// tuples of (attributed breakdown key, attributed trigger value) for each
+// record.
+// The output is a Histogram. BKs are assigned by the advertiser and sent in
+// the input of IPA. BK values are expected to be dense.
+// How breakdown keys are defined is out-of-scope.
+//
+// High level explanation of the protocol:
+// 1. TODO: Add fake attribution outputs.
+// 2. Shuffle.
+// 3. Reveal breakdown. Trigger values are not revelaed.
+// 4. Aggregation if trigger values secret shares.
```

Comment:
Yeah, I still need to come up with a way to chunk/vectorize things.

Reply:
don't we know the breakdown in the clear, so we can trivially group them as we reveal it?

Reply:
yes its all local so you could sort by breakdown and then vectorize within events with the same breakdown. Sorting a large list might still cause a lot of reallocations which might not be necessary.

Reply:
Yeah, for the breakdown counts that we're contemplating currently, I would say just fan the values out into a bucket per breakdown. Then once we get two chunks worth of records in a bucket, we can do a vectorized reduction.

If we're doing a full resharding here (which a sharded shuffle is), then maybe it makes sense to shard by breakdowns rather than randomly? There is a question whether we get better parallelism across breakdowns or by eagerly pulling more chunks from a particular aggregation stream.

Reply:
> Yes its all local so you could sort by breakdown
I still don't see a need to sort a big list by breakdown key. As Andy pointed out, we fan out events to the corresponding shards by the breakdown key as we drain the input. Grouping comes essentially for free

> If we're doing a full resharding here (which a sharded shuffle is), then maybe it makes sense to shard by breakdowns rather than randomly? 

This makes sense to me - maybe in sharded world in this case we can avoid transpose 

> There is a question whether we get better parallelism across breakdowns or by eagerly pulling more chunks from a particular aggregation stream.

To make a decision here, it would be helpful to understand the distribution of events per breakdown. If it is not close to uniform in presence of fake events (basically follows user-provided distribution), then we can assume the worst case and vectorize across breakdowns per shard